Feature/move agent ucc listener to backhaul manager

morantr opened the pull request:
move agent ucc listener to backhaul manager and remove it from the son_slave.

tomereli commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -77,6 +77,10 @@ backhaul_manager::backhaul_manager(const config_file::sConfigSlave &config,
     configuration_stop_on_failure_attempts = string_utils::stoi(config.stop_on_failure_attempts);
     stop_on_failure_attempts               = configuration_stop_on_failure_attempts;
     LOG(DEBUG) << "stop_on_failure_attempts=" << configuration_stop_on_failure_attempts;
+    m_sConfig.ucc_listener_port = string_utils::stoi(config.ucc_listener_port);
```
I think that all certification related members should be moved to a separate class, and the backhaul manager thread should have a member of that class which we can compile out when the time comes and we are product ready.
For the same reason, it makes sense to have the certification parameters in a separate configuration file, not part of the agent configuration file so it won't be accidently moved to the platform DB when we implement #568.

morantr replied:
I put it there since all the configurations from the configuration file are there. 
We still can compile it out with `#ifdef` where it currently placed.
I don't understand why it can't be moved to DB with all other configurations. Why should it be different?

tomereli replied:
If we can make an optional section in the DB, something like:
```
prplmesh.config.certification
   enable
   port
```
Then fine, I guess it can work. anyway, (nitpick), should be done as part of #575

tomereli commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.h:
```diff
@@ -120,6 +120,9 @@ class backhaul_manager : public btl::transport_socket_thread {
         std::string wire_iface;
         std::string wireless_iface;
         std::string preferred_bssid;
+        std::string vendor;
```
Same here, should be a separate class / structure.

tomereli commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.h:
```diff
@@ -35,7 +35,9 @@ class backhaul_manager : public btl::transport_socket_thread {
 
-protected:
+    // For agent_ucc_listener
+    const std::string &get_controller_bridge_mac() { return controller_bridge_mac; }
```
I think this commit is redundant. We are in control path, there is nothing wrong with doing a system call whenever we need the controller bridge MAC. If we do that every time we need it, we are also adding the infrastructure for being able to cope with dynamic changes to the main bridge, which is hardcoded everywhere currently.

morantr replied:
But there is no system call to get the remote controller bridge mac.
Only the backhaul manager holds it, and it gets it after receiving an auto-configuration search message response.
😕

tomereli replied:
Then the name should change to `controller_almac`

tomereli commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -286,6 +286,9 @@ void backhaul_manager::platform_notify_error(int code, const std::string &error_
 }
 
+void backhaul_manager::before_select()
+{
```
How can this compile??? you're missing a `}`

morantr replied:
It was added on later commit. Guess I forgot to git add the close bracket on that specific commit and added it on one of the following commits. Anyway, fixed.

tomereli commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -286,6 +286,9 @@ void backhaul_manager::platform_notify_error(int code, const std::string &error_
     message_com::send_cmdu(m_scPlatform.get(), cmdu_tx);
 }
 
+void backhaul_manager::before_select()
```
commit message typo - "functiob" --> "function".
Can you also please elaborate a bit more on the use of the mutex in this commit message? I still don't understand it completely..

morantr replied:
fixed

tomereli commented on agent/config/beerocks_agent.conf.in:
```diff
@@ -12,7 +12,6 @@ temp_path=@TMP_PATH@
 vendor=Intel
 model=prplMesh
 ucc_listener_port=8002 # 0 - disabled
-ucc_listener_slave_hostap_iface=wlan0
```
commit message grammar - " will be move " should be "will move" or "will be moved"

tomereli replied:
Also, this should have been done as part of the commit which removed the previous implementation, shouldn't it?

morantr replied:
fixed the grammar mistake.
The previous commit removed a completely different thing, so no.

tomereli commented on agent/src/beerocks/slave/agent_ucc_listener.h:
```diff
@@ -15,10 +15,15 @@
 public:
-    agent_ucc_listener(uint16_t port, const std::string &vendor, const std::string &model,
-                       const std::string &bridge_iface, SocketClient **controller_sd);
+    agent_ucc_listener(backhaul_manager &backhaul_manager_ctx, uint16_t port,
```
Shouldn't this be `const backhaul_manager &backhaul_manager_ctx`?

tomereli replied:
Isn't it possible to make the `send_cmdu_to_bus` a friend function to the `agent_ucc_listener` instead of saving the full context?

morantr replied:
It can't because passing `this` pointer discards the `const` qualifier.
Either way, `friend` won't help since we need the instance of the backhaul manager, which we can get only by passing the 'this' pointer.
`friend` could help if we want to access private members. For example, instead of creating function to get `controller_bridge _mac` as I did in one of the previous commits, it could directly get access to it. I have chosen to create a function for that since that way it could not modify or change anything. This gives more like `const` behavior without actually declaring it as `const`.

tomereli commented on agent/src/beerocks/slave/agent_ucc_listener.cpp:
```diff
@@ -75,7 +75,7 @@ bool agent_ucc_listener::validate_destination_alid(const std::string &dest_alid)
     // On the agent side, the dest_alid is not really needed since the destination socket will
     // always be the controller socket.
-    return (*m_controller_sd)->getPeerMac() == dest_alid;
+    return m_backhaul_manager_ctx.get_controller_bridge_mac() == dest_alid;
```
This shouldn't be using the bridge mac. We should add a confiruation parameter to the agent which is al_mac, and for now set it to the same address as the bridge. Or at least call it here and everywhere we use it al_mac and not controller_bridge_mac, since in the future this might change.

morantr replied:
Since we work now with bridge mac, I prefer to use `bridge` terminology to prevent confusion. When we will the bridge mac will stop from being the bridge mac we will change it.

arnout replied:
I think this proves that we should clean up our terminology sooner rather than later :-)

Since currently indeed controller_bridge_mac is used in a few places already, I agree that it's better to use the same term in new code in backhaul_manager and son_slave_thread, even if it is wrong.

tomereli replied:
We can at least add a comment in those places saying its the al_mac..

tomereli commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -120,6 +120,16 @@ bool backhaul_manager::init()
         return false;
     }
 
+    if (m_sConfig.ucc_listener_port != 0) {
```
We should add a sanity test which verifies the selected port is a valid port which is not used by another application (nitpick)

morantr replied:
I think there is, and it is done by the socket thread. The bind function will fail on socket thread if someone using it.

tomereli commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -120,6 +120,16 @@ bool backhaul_manager::init()
+        m_agent_ucc_listener = std::make_unique<agent_ucc_listener>(
+            *this, m_sConfig.ucc_listener_port, m_sConfig.vendor, m_sConfig.model,
+            m_sConfig.bridge_iface);
+        if (m_agent_ucc_listener && !m_agent_ucc_listener->start("ucc_listener")) {
```
perhaps the name should be `prplmesh_agent_ucc_listener`?

morantr replied:
It is for logging purposes, so I think it is better to use the shortest name possible. Some places leave it empty...

tomereli commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -120,6 +120,16 @@ bool backhaul_manager::init()
+            *this, m_sConfig.ucc_listener_port, m_sConfig.vendor, m_sConfig.model,
+            m_sConfig.bridge_iface);
+        if (m_agent_ucc_listener && !m_agent_ucc_listener->start("ucc_listener")) {
+            LOG(ERROR) << "failed start agent_ucc_listener";
```
We should print the error but not fail the whole agent init on this, IMO.

morantr replied:
I disagree. If an error occurs, there is no way to continue certification. So it is better to fail the init function so we won't be dragged into countless backhaul resets if something will go wrong (and it will).
Also, if the failure is due to the memory allocation failure, this alone should be failing the init.

tomereli commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -288,9 +298,17 @@ void backhaul_manager::platform_notify_error(int code, const std::string &error_
 void backhaul_manager::before_select()
 {
+    if (m_agent_ucc_listener) {
+        m_agent_ucc_listener->unlock();
```
This deserves an explanation in the commit message and a comment which explains exactly what the `lock` and `unlock` are protecting.

morantr replied:
I explained it in the past, when I developed the ucc_listener.
It is needed since the ucc_listener runs in a different thread and shares the same memory with the backhaul manager.

tomereli commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.h:
```diff
@@ -218,6 +221,8 @@ class backhaul_manager : public btl::transport_socket_thread {
     std::string controller_bridge_mac;
     std::string bssid_bridge_mac;
 
+    std::unique_ptr<beerocks::agent_ucc_listener> m_agent_ucc_listener;
```
Why does it need to be a unique_ptr? it makes more since to add this as a reference which can be wrapped with an `#ifdef CERTIFICATION_SUPPORT` in the future.

morantr replied:
Because reference needs to be initialized in the constructor, and I wanted to construct the ucc_listener on the backhaul manager init function, so it would fail the init function if something goes wrong. I used the same design for the implementation in the controller, and it was the same way when it was implemented on the son_slave.
Either way, it is still possible to wrap it around with `#ifdef`, just wrap a pointer instead of reference. There is no difference.

tomereli reviewed (changes requested):
Apart from the inline comments, this PR is lacking almost all the functionality added by #531:
- `dev_reset_defaults` should do a full agent reset. The backhaul FSM should stop before searching the controller and wait for `dev_set_config,name,IntelCTTDUT,program,map,backhaul,eth` command which only then should continue with autoconfig search.
- `dev_get_param` missing support for `macaddr` parameter

Relevant commits from #531:
https://github.com/prplfoundation/prplMesh/pull/531/commits/a5fd712982ad2b7c3c54cc5ad01940d432e64d65
https://github.com/prplfoundation/prplMesh/pull/531/commits/bcdcac0eee234bb7c1f1098f05c02f79d2b8ad27

Then you will also need the VAPs list update to be saved in the agent, so we can really answer dev_get_config correctly, but this should be done in a followup PR. For now just return ERROR.

morantr committed:
agent: Remove agent_ucc_listener from son_slave

Preparative commit.
Remove agent_ucc_listener from son_slave in order to move it to the
backhaul manager:
Removed agent_ucc_listener member, initialization and
"before/after_select" functions that was overridden for implementing it
in the son_slave.

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

morantr committed:
agent: Remove "backhaul_manager" namespace

The class name of the backhaul manager was "main_thread" and it was
wrapped with "backhaul_manager" namespace.
There is no real good reason to do it, and it is better to call the
backhaul_manager thread "backhaul_manager"  instead of "main_thread".
Remove "backhaul_manager" namespace, and change the name of the
backhaul manager thread to "backhaul_manager".

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

morantr committed:
backhaul_mananger: Add "vendor", "model" and "ucc_listener_port" to conf

Preparative commit.
Add "vendor", "model" and "ucc_listener_port" from the agent
the configuration file to the backhaul configuration struct and save them
on the backhaul manager constructor.
As well save the "bridge_iface" parameter form agent configuration file
to the backhaul constructor.
These changes are needed for the initialization of the
agent_ucc_listener on the backhaul manager.

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

morantr committed:
backhaul_manager: Add "get_controller_bridge_mac" function

Preparative commit.
Add "get_controller_bridge_mac" function which will be in use of the
agent_ucc_listener.

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

morantr committed:
backhaul_mananger:  Add "before_select" function override

Preparative commit.
Add "before_select" function override to backhaul manager.
In the follow up commit this function will be used to unlock the mutex
of the ucc_listener.
In addition, move a bunch of functions on the backhaul manager to be
private instead of protected, since no class should be derived from the
backhaul manager anyway.

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

morantr committed:
backhaul_manager: Remove "bridge_iface" from cACTION_BACKHAUL_ENABLE

Now that "bridge_iface" is being set on the backhaul manager
constructor, there is no need to pass and set it upon receiving
ACTION_BACKHAUL_ENABLE message, so it has been removed from it.

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

morantr committed:
agent: Remove "ucc_listener_slave_hostap_iface" from configuration file

Preparative commit.
Since the agent_ucc_listener will be moved from the son_slave to the
backhaul manager in the followup commit, there is no need to hold on
which son_slave should run the ucc_listener anymore.
Remove "ucc_listener_slave_hostap_iface" from configuration file.

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

morantr committed:
agent_ucc_listener: Add backhaul manager context

Add backhaul manager context to agent_ucc_listener which is necessary
for getting access to backhaul manager "send_cmdu_to_bus" and
"get_controller_bridge_mac" functions.
Also, remove controller socket since it is not relevant anymore.

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

morantr committed:
agent_ucc_listener: Update implementation

Update the implementation of "send_cmdu_to_destination" and
"validate_destination_alid" function to adapt to the backhaul manager
instead of the son_slave.

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

morantr committed:
agent_ucc_listener: Fix wrong "m_ucc_listener_run_on" initialization

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

morantr committed:
backhaul_manager: Add the agent_ucc_listener

Add the agent_ucc_listener to the backhaul manager:
agent_ucc_listener construction on the backhaul manager init function
and add ucc_listener mutex locking on "before/after_select" functions.

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

tomereli commented:
![](https://github.trello.services/images/mini-trello-icon.png) [agent ucc listener refactor](https://trello.com/c/paMpixDg/7-agent-ucc-listener-refactor)

arnout commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.h:
```diff
@@ -11,6 +11,7 @@
 
 #include "wan_monitor.h"
 
+#include <bcl/beerocks_backport.h>
```
make_unique is not called in this header, it's called in backhaul_manager.cpp, so the include should be there.
